### ISSUE TITLE
MuonGun Track.Harvest bug (Trac #1583)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1575]
Energy deposited at some point along the track sometimes returns nonsensical negative energy. File below (located at DESY Zeuthen) contains I3MCTree and MMCTrackList for several example events. 

/lustre/fs11/group/icecube/blotsd/muon_gun/muon_gun_testcase.i3.bz2

Relevant information about the file:

**Generation**
model.flux.max_multiplicity = 1
spectrum = MuonGun.OffsetPowerLaw(5,100*I3Units.TeV, 1*I3Units.GeV,1*I3Units.TeV)
outer = MuonGun.Cylinder(1600*I3Units.m, 800*I3Units.m)
model = MuonGun.load_model('GaisserH4a_atmod12_SIBYLL')
generator = MuonGun.Floodlight(outer, spectrum, 0.99,1.0)

***Propagation***
muprop = I3PropagatorServicePROPOSAL(type=dataclasses.I3Particle.MuMinus, cylinderHeight=1600, cylinderRadius=800)

If I try to get the deposited energy in the volume as below sometimes its negative.

for track in MuonGun.Track.harvest(frame['I3MCTree'], frame['MMCTrackList']):
    # Find distance to entrance and exit from sampling volume
    intersections = outer.intersection(track.pos, track.dir)
    # Get the corresponding energies
    e0, e1 = track.get_energy(intersections.first)/I3Units.GeV, track.get_energy(intersections.second)/I3Units.GeV

Sometimes  e1 < 0  
and that means that e0 - e1 > 1 TeV (maximum generated energy)

Migrated from https://code.icecube.wisc.edu/ticket/1583
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:13:10", 
    "description": "Energy deposited at some point along the track sometimes returns nonsensical negative energy. File below (located at DESY Zeuthen) contains I3MCTree and MMCTrackList for several example events. \n\n/lustre/fs11/group/icecube/blotsd/muon_gun/muon_gun_testcase.i3.bz2\n\nRelevant information about the file:\n\n**Generation**\nmodel.flux.max_multiplicity = 1\nspectrum = MuonGun.OffsetPowerLaw(5,100*I3Units.TeV, 1*I3Units.GeV,1*I3Units.TeV)\nouter = MuonGun.Cylinder(1600*I3Units.m, 800*I3Units.m)\nmodel = MuonGun.load_model('GaisserH4a_atmod12_SIBYLL')\ngenerator = MuonGun.Floodlight(outer, spectrum, 0.99,1.0)\n\n***Propagation***\nmuprop = I3PropagatorServicePROPOSAL(type=dataclasses.I3Particle.MuMinus, cylinderHeight=1600, cylinderRadius=800)\n\n\nIf I try to get the deposited energy in the volume as below sometimes its negative.\n\nfor track in MuonGun.Track.harvest(frame['I3MCTree'], frame['MMCTrackList']):\n    # Find distance to entrance and exit from sampling volume\n    intersections = outer.intersection(track.pos, track.dir)\n    # Get the corresponding energies\n    e0, e1 = track.get_energy(intersections.first)/I3Units.GeV, track.get_energy(intersections.second)/I3Units.GeV\n\nSometimes  e1 < 0  \nand that means that e0 - e1 > 1 TeV (maximum generated energy)\n", 
    "reporter": "summer.blot", 
    "cc": "thomas.kintscher@desy.de", 
    "resolution": "fixed", 
    "_ts": "1550067190995086", 
    "component": "combo simulation", 
    "summary": "MuonGun Track.Harvest bug", 
    "priority": "normal", 
    "keywords": "MuonGun", 
    "time": "2016-03-10T13:44:14", 
    "milestone": "", 
    "owner": "jvansanten", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20160310T15:08:16: 
* jvansanten commented:

> In changeset 142926:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="142926"
> Fix an issue where energy losses were double-counted between the last stochastic and the border of the MMC volume. Fixes #1575.
> ```


* jvansanten changed resolution from "" to "fixed"
* jvansanten changed status from "new" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20160523T20:03:26: 
* thomas.kintscher changed resolution from "fixed" to ""
* thomas.kintscher changed status from "closed" to "reopened"
* thomas.kintscher changed cc from "" to "thomas.kintscher@desy.de"
* thomas.kintscher commented:

> The assertion in the bugfix trips when running the code snippet from the original ticket (which originates from the [http://software.icecube.wisc.edu/documentation/projects/MuonGun/utilities.html docs]) on default simulation datasets.
> 
> Steps to reproduce:
> - Open for example this file in pyshovel:
> 
> ```text
> dataio-pyshovel /data/sim/IceCube/2012/filtered/level2/neutrino-generator/11070/00000-00999/Level2_nugen_numu_IC86.2012.011070.000433.i3.bz2
> ```
> - Go to frame 4.
> - Switch to the iPython console and run:
> 
> ```text
> from icecube import MuonGun, phys_services
> surf = phys_services.Cylinder(1000., 500.)
> edep = 0.
> for track in MuonGun.Track.harvest(frame['I3MCTree'], frame['MMCTrackList']):
>     intersect = surf.intersection(track.pos, track.dir)
>     e0 = track.get_energy(intersect.first)
>     e1 = track.get_energy(intersect.second)
>     edep += (e0 - e1)
> ```
> - End up with:
> 
> ```text
> FATAL (MuonGun): Assertion failed: ls->energy <= cp->energy && "sum of losses is smaller than energy at last checkpoint" (Track.cxx:85 in double I3MuonGun::Track::GetEnergy(double) const)
> ```
> 
> It's not clear to me what the solution to this is, and if it can be avoided before calling the method in question. Would it be possible to fail (more) gracefully here?

**Comment by icecube-trac** [comment]
Trac update at 20160524T08:20:12: jvansanten commented:

> This points to a deeper bug where the CMC expansion of the hadronic cascade was attached to a Michel electron in the decompression step:
> 
> 
> ```text
>   456 Hadrons (-314.836m ,359.404m ,486.365m) (41.1846deg ,300.289deg) 9648.23ns 6.55012e+06GeV 0m 
>     4899 MuPlus (-314.836m ,359.404m ,486.365m) (41.1846deg ,300.289deg) 9648.23ns 106.303GeV 422.445m 
>       6762 EPlus (-455.136m ,599.606m ,168.436m) (41.1846deg ,300.289deg) 15081ns 0.0340286GeV 0m 
>     4900 MuPlus (-314.836m ,359.404m ,486.365m) (41.1846deg ,300.289deg) 9648.23ns 74.4544GeV 314.318m 
>       6767 EPlus (-419.225m ,538.125m ,249.812m) (41.1846deg ,300.289deg) 13199.2ns 0.0279383GeV 0m 
>     4901 MuPlus (-314.836m ,359.404m ,486.365m) (41.1846deg ,300.289deg) 9648.23ns 52.1812GeV 205.792m 
>       6747 EMinus (-314.836m ,359.404m ,486.365m) (41.1846deg ,300.289deg) 9648.23ns 53.1932GeV 0m 
>       6748 EMinus (-315.194m ,360.018m ,485.552m) (41.1846deg ,300.289deg) 9651.83ns 989.082GeV 0m 
>       6749 EMinus (-315.553m ,360.632m ,484.739m) (41.1846deg ,300.289deg) 9655.43ns 2613.05GeV 0m 
>       6750 EMinus (-315.912m ,361.246m ,483.927m) (41.1846deg ,300.289deg) 9659.03ns 1993.01GeV 0m 
>       6773 EPlus (-383.182m ,476.417m ,331.488m) (41.1846deg ,300.289deg) 11337.3ns 0.0468289GeV 0m 
>         4909 EMinus (-383.182m ,476.417m ,331.488m) (41.1846deg ,300.289deg) 11337.3ns 44.4149GeV 0m 
>         4910 EMinus (-383.541m ,477.031m ,330.675m) (41.1846deg ,300.289deg) 11340.9ns 12274.3GeV 0m 
>         4911 EMinus (-383.9m ,477.645m ,329.862m) (41.1846deg ,300.289deg) 11344.5ns 174424GeV 0m 
>         4912 EMinus (-384.258m ,478.259m ,329.049m) (41.1846deg ,300.289deg) 11348.1ns 673303GeV 0m 
>         4913 EMinus (-384.617m ,478.873m ,328.237m) (41.1846deg ,300.289deg) 11351.7ns 1.23851e+06GeV 0m 
>         4914 EMinus (-384.976m ,479.487m ,327.424m) (41.1846deg ,300.289deg) 11355.3ns 1.41977e+06GeV 0m 
>         4915 EMinus (-385.334m ,480.101m ,326.611m) (41.1846deg ,300.289deg) 11358.9ns 1.17524e+06GeV 0m 
>         4916 EMinus (-385.693m ,480.716m ,325.798m) (41.1846deg ,300.289deg) 11362.5ns 767551GeV 0m 
>         4917 EMinus (-386.052m ,481.33m ,324.985m) (41.1846deg ,300.289deg) 11366.1ns 418759GeV 0m 
>         4918 EMinus (-386.41m ,481.944m ,324.173m) (41.1846deg ,300.289deg) 11369.7ns 198413GeV 0m 
>         4919 EMinus (-386.769m ,482.558m ,323.36m) (41.1846deg ,300.289deg) 11373.3ns 83925.8GeV 0m 
>         4920 EMinus (-387.128m ,483.172m ,322.547m) (41.1846deg ,300.289deg) 11376.9ns 32339.6GeV 0m 
>         4921 EMinus (-387.486m ,483.786m ,321.734m) (41.1846deg ,300.289deg) 11380.5ns 11527.3GeV 0m 
>         4922 EMinus (-387.845m ,484.4m ,320.921m) (41.1846deg ,300.289deg) 11384.1ns 3846.05GeV 0m 
> ```
> 
> Will poke further.

**Comment by icecube-trac** [comment]
Trac update at 20160524T15:10:58: olivas changed milestone from "" to "July 2016"

**Comment by icecube-trac** [comment]
Trac update at 20160525T17:49:26: 
* jvansanten commented:

> This tree is correctly reassembled in the simulation release it was made with (04-01-07). The problem is related to the new hash-based I3MCTree. One clue is that the order of the primaries in this event is different in 04-01-07:
> 
> 
> ```text
> [icecube.dataclasses.ParticleType.PPlus, icecube.dataclasses.ParticleType.Mg24Nucleus, icecube.dataclasses.ParticleType.NuMu]
> ```
> 
> and current dataclasses:
> 
> 
> ```text
> [icecube.dataclasses.ParticleType.PPlus, icecube.dataclasses.ParticleType.NuMu, icecube.dataclasses.ParticleType.Mg24Nucleus]
> ```
> 
> Since I3LinearizedMCTree traverses the tree in pre-order to reattach the leaves, this is the likely cause of the scramble.


* jvansanten changed _comment0 from:

> This tree is correctly reassembled in the simulation release it was made with (04-01-07). The problem is related to the new hash-based I3MCTree. One clue is that the order of the primaries in this event is different in 04-01-07:
> 
> {{{
> [icecube.dataclasses.ParticleType.PPlus, icecube.dataclasses.ParticleType.NuMu, icecube.dataclasses.ParticleType.Mg24Nucleus]
> }}}
> 
> and current dataclasses:
> 
> {{{
> [icecube.dataclasses.ParticleType.PPlus, icecube.dataclasses.ParticleType.Mg24Nucleus, icecube.dataclasses.ParticleType.NuMu]
> }}}
> 
> Since I3LinearizedMCTree traverses the tree in pre-order to reattach the leaves, this is the likely cause of the scramble.

to:

> 1464199367010088

**Comment by icecube-trac** [comment]
Trac update at 20160525T18:17:38: 
* jvansanten commented:

> In changeset 146403:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="146403"
> Insert primaries from old-style trees in order, rather than putting the
> first one first, then reversing the rest.  Fixes #1575.
> 
> This is depressingly similar to r104860.
> ```


* jvansanten changed resolution from "" to "fixed"
* jvansanten changed status from "reopened" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20160704T14:43:27: 
* thomas.kintscher commented:

> Thank you, this resolved most of the problems, especially the one mentioned above!
> 
> However, with the same kind of simulation, it's still possible to trip on other assertion. It's not clear whether this is related as I have trouble identifying the line in the code.
> However, I didn't see this before doing the ```svn up``` in the MuonGun, i.e. before r146403.
> 
> To reproduce, open this file:
> 
> ```text
> dataio-pyshovel /data/sim/IceCube/2012/filtered/level2/neutrino-generator/11069/01000-01999//Level2_nugen_numu_IC86.2012.011069.001612.i3.bz2
> ```
> 
> Go to event 1287.
> 
> Execute this:
> 
> ```text
> In [1]: from icecube import MuonGun
> 
> In [2]: tracks = MuonGun.Track.harvest(frame['I3MCTree'], frame['MMCTrackList'])
> FATAL (Tree): Assertion failed: storage_ != treePtr_->end_ (I3MCTree.h:221 in ValueType& TreeBase::Tree<T, Key, Hash>::iterator_storage_impl<boost::optional<Key>, ValueType, TreeType, TreeHashMapIter>::dereference() const [with ValueType = const I3Particle, TreeType = const TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> >, TreeHashMapIter = __gnu_cxx::_Hashtable_const_iterator<std::pair<const I3ParticleID, TreeBase::TreeNode<I3Particle> >, I3ParticleID, __gnu_cxx::hash<I3ParticleID>, std::_Select1st<std::pair<const I3ParticleID, TreeBase::TreeNode<I3Particle> > >, std::equal_to<I3ParticleID>, std::allocator<TreeBase::TreeNode<I3Particle> > >, T = I3Particle, Key = I3ParticleID, Hash = __gnu_cxx::hash<I3ParticleID>])
> ---------------------------------------------------------------------------
> RuntimeError                              Traceback (most recent call last)
> <ipython-input-2-c16f962d940c> in <module>()
> ----> 1 tracks = MuonGun.Track.harvest(frame['I3MCTree'], frame['MMCTrackList'])
> 
> RuntimeError: Assertion failed: storage_ != treePtr_->end_ (in ValueType& TreeBase::Tree<T, Key, Hash>::iterator_storage_impl<boost::optional<Key>, ValueType, TreeType, TreeHashMapIter>::dereference() const [with ValueType = const I3Particle, TreeType = const TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> >, TreeHashMapIter = __gnu_cxx::_Hashtable_const_iterator<std::pair<const I3ParticleID, TreeBase::TreeNode<I3Particle> >, I3ParticleID, __gnu_cxx::hash<I3ParticleID>, std::_Select1st<std::pair<const I3ParticleID, TreeBase::TreeNode<I3Particle> > >, std::equal_to<I3ParticleID>, std::allocator<TreeBase::TreeNode<I3Particle> > >, T = I3Particle, Key = I3ParticleID, Hash = __gnu_cxx::hash<I3ParticleID>])
> ```
> 
> 


* thomas.kintscher changed resolution from "fixed" to ""
* thomas.kintscher changed status from "closed" to "reopened"

**Comment by icecube-trac** [comment]
Trac update at 20160704T19:06:15: thomas.kintscher commented:

> I have created a minimal example which triggers the error here:
> ```/data/user/tkintscher/bugs/muongun_assertion.py```
> 
> Running it in gdb with an IceRec/trunk debug build yields this backtrace:
> 
> ```text
> FATAL (Tree): Assertion failed: storage_ != treePtr_->end_ (I3MCTree.h:221 in ValueType& TreeBase::Tree<T, Key, Hash>::iterator_storage_impl<boost::optional<Key>, ValueType, TreeType, TreeHashMapIter>::dereference() const [with ValueType = const I3Particle, TreeType = const TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> >, TreeHashMapIter = __gnu_cxx::_Hashtable_const_iterator<std::pair<const I3ParticleID, TreeBase::TreeNode<I3Particle> >, I3ParticleID, __gnu_cxx::hash<I3ParticleID>, std::_Select1st<std::pair<const I3ParticleID, TreeBase::TreeNode<I3Particle> > >, std::equal_to<I3ParticleID>, std::allocator<TreeBase::TreeNode<I3Particle> > >, T = I3Particle, Key = I3ParticleID, Hash = __gnu_cxx::hash<I3ParticleID>])
> Catchpoint 1 (exception thrown), 0x00007ffff0687cb0 in __cxa_throw () from /usr/lib64/libstdc++.so.6
> Missing separate debuginfos, use: debuginfo-install bzip2-libs-1.0.5-7.el6_0.x86_64 freetype-2.3.11-15.el6_6.1.x86_64 keyutils-libs-1.4-4.el6.x86_64 krb5-libs-1.10.3-42z1.el6_7.x86_64 libcom_err-1.42.7.wc2-7.el6.x86_64 libgcc-4.4.7-3.el6.x86_64 libgfortran-4.4.7-3.el6.x86_64 libselinux-2.0.94-5.3.el6_4.1.x86_64 libstdc++-4.4.7-3.el6.x86_64 openssl-1.0.1e-48.sl6_8.1.x86_64 zlib-1.2.3-29.el6.x86_64
> (gdb) where
> https://code.icecube.wisc.edu/ticket/0  0x00007ffff0687cb0 in __cxa_throw () from /usr/lib64/libstdc++.so.6
> #1  0x00007fffe6a3db99 in TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> >::iterator_storage_impl<boost::optional<I3ParticleID>, I3Particle const, TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> > const, __gnu_cxx::_Hashtable_const_iterator<std::pair<I3ParticleID const, TreeBase::TreeNode<I3Particle> >, I3ParticleID, __gnu_cxx::hash<I3ParticleID>, std::_Select1st<std::pair<I3ParticleID const, TreeBase::TreeNode<I3Particle> > >, std::equal_to<I3ParticleID>, std::allocator<TreeBase::TreeNode<I3Particle> > > >::dereference (this=0x7fffffffbe70) at /data/user/tkintscher/software/icerec-trunk/src/dataclasses/public/dataclasses/physics/I3MCTree.h:221
> #2  0x00007fffe6a3d6f4 in TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> >::iterator_base<TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> >::pre_order<I3Particle const>, I3Particle const, boost::optional<I3ParticleID> >::dereference (
>     this=0x7fffffffbe70) at /data/user/tkintscher/software/icerec-trunk/src/dataclasses/public/dataclasses/physics/I3MCTree.h:378
> #3  0x00007fffe6a3cef6 in boost::iterators::iterator_core_access::dereference<TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> >::iterator_base<TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> >::pre_order<I3Particle const>, I3Particle const, boost::optional<I3ParticleID> > > (f=...) at /cvmfs/icecube.opensciencegrid.org/py2-v2/RHEL_6_x86_64/include/boost/iterator/iterator_facade.hpp:540
> #4  0x00007fffe6a3bd24 in boost::iterators::detail::iterator_facade_base<TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> >::iterator_base<TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> >::pre_order<I3Particle const>, I3Particle const, boost::optional<I3ParticleID> >, I3Particle const, boost::iterators::forward_traversal_tag, I3Particle const&, long, false, false>::operator* (this=0x7fffffffbe70) at /cvmfs/icecube.opensciencegrid.org/py2-v2/RHEL_6_x86_64/include/boost/iterator/iterator_facade.hpp:646
> #5  0x00007fffd3e4faf0 in I3MuonGun::Track::Harvest (mctree=..., mmctracks=...) at /data/user/tkintscher/software/icerec-trunk/src/MuonGun/private/MuonGun/Track.cxx:151
> ```

**Comment by icecube-trac** [comment]
Trac update at 20160729T08:00:19: jvansanten commented:

> Replying to [comment:8 thomas.kintscher]:
> 
> The event in question looks like this:
> 
> ```text
> 175003 NuMu (-2.52471e+06m ,2.23581e+06m ,-963251m) (105.933deg ,138.471deg) -1.16901e+07ns 242.808GeV 3.5075e+06m 
>   2 MuMinus (190.878m ,-336.748m ,-401.468m) (105.933deg ,138.471deg) 9682.06ns 0.105658GeV 0m 
>     175007 EMinus (190.878m ,-336.748m ,-401.468m) (105.933deg ,138.471deg) 13735.6ns 0.0399987GeV 0m 
>   3 Hadrons (190.878m ,-336.748m ,-401.468m) (105.933deg ,138.471deg) 9682.06ns 242.702GeV 0m 
>     175005 EMinus (190.878m ,-336.748m ,-401.468m) (105.933deg ,138.471deg) 9682.06ns 219.083GeV 0m 
> ```
> The muon in question has a total energy of 105 MeV; in other words it is at rest. It was marked with shape `Dark`, meaning it should not emit light. The other major user of the `Dark` shape is CMC; in old releases it re-used particle IDs, which breaks the new MCTree. This use of Dark caused the new MCTree to assign a new particle ID that was different from the one in the MMCTrackList. Ugly.

**Comment by icecube-trac** [comment]
Trac update at 20160729T08:03:27: 
* jvansanten changed resolution from "" to "fixed"
* jvansanten commented:

> In changeset 148886:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="148886"
> If a particle is marked Dark and stored in an old-style MCTree, the
> new-style MCTree will give it a new event ID on the assumption that
> it came from CMC. Muons with energies equal to the muon rest mass are
> apparently also marked Dark, which trips the same mechanism. To account
> for this, check the result of the particle-ID search before dereferencing
> it. Fixes #1575 (again).
> ```


* jvansanten changed status from "reopened" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:13:10: 
* olivas commented:

> Milestone July 2016 deleted


* olivas changed milestone from "July 2016" to ""
